RickAndMortyFunApp/NetworkServices/NetworkService.swift

```diff
@@ -6,56 +6,97 @@
+
+    case get(Int)
+    case create([String: Any])
+    case delete(Int)
+
+    func asURLRequest() throws -> URLRequest {
+        var method: HTTPMethod {
+            switch self {
+            case .get:
+                return .get
+            case .create:
+                return .post
+            case .delete:
+                return .delete
+            }
+        }
+
+        let params: ([String: Any]?) = {
+            switch self {
+            case .get, .delete:
+                return nil
+            case .create(let newTodo):
+                return (newTodo)
+            }
+        }()
+
+        let url: URL = {
+            var components = URLComponents(string: NetworkRouter.baseURLString)
+            switch self {
+            case .get(let page):
+                components?.queryItems = [
+                    URLQueryItem(name: "page", value: String(page))
+                ]
+            case .create:
+                components?.query = ""
+            case .delete:
+                components?.query = ""
+            }
+            var url = URL(string: NetworkRouter.baseURLString)!
+            if let urlWithComponents = components?.url {
+                url = urlWithComponents
+            }
+            return url
+        }()
+
+        var urlRequest = URLRequest(url: url)
+        urlRequest.httpMethod = method.rawValue
+
+        let encoding: ParameterEncoding = {
+            switch method {
+            case .get:
+                return  URLEncoding.default
+            default:
+                return JSONEncoding.default
+            }
+        }()
+        return try encoding.encode(urlRequest, with: params)
+    }
+}
+
 class NetworkService {
 
-    let urlString = "https://rickandmortyapi.com/api/character/"
-    var task: URLSessionTask?
-    
     func loadCharcters(matching query: String?, page: Int, completion: @escaping(CharactersResult) -> Void) {
 
         func fireErrorCompletion(_ error: Error?) {
             completion(CharactersResult(characters: nil, error: error,
                                         currentPage: 0, pageCount: 0))
         }
-
-        var queryOrEmpty = "since:1970-01-02"
-        if let query = query, !query.isEmpty {
-            queryOrEmpty = query
-        }
-        var components = URLComponents(string: urlString)
-        components?.queryItems = [
-            URLQueryItem(name: "query", value: queryOrEmpty),
-            URLQueryItem(name: "page", value: String(page))
-        ]
-        guard let url = components?.url else {
-            fireErrorCompletion(NetworkError.invalidURL)
-            return
-        }
-        task?.cancel()
-
-        task = URLSession.shared.dataTask(with: url) { (data, respone, error) in
-            DispatchQueue.main.async {
-                guard let data = data else {
-                    fireErrorCompletion(error)
+        AF.request(NetworkRouter.get(page)).responseDecodable( queue: nil , decoder: JSONDecoder() )
```

Comment:
JSONDecoder() is the default parameter. Look inside this function. So you can remove it from this line

Reply:
queue: nil is also the default value